go/pools/smartconnpool/pool.go

```diff
@@ -0,0 +1,706 @@
+}
+
+func (pool *ConnPool[C]) reopen() {
+	capacity := pool.capacity.Load()
+	if capacity == 0 {
+		return
+	}
+
+	pool.Close()
+	pool.open()
+	pool.SetCapacity(capacity)
+}
+
+// IsOpen returns whether the pool is open
+func (pool *ConnPool[C]) IsOpen() bool {
+	return pool.close != nil
+}
+
+// Capacity returns the maximum amount of connections that this pool can maintain open
+func (pool *ConnPool[C]) Capacity() int64 {
+	return pool.capacity.Load()
+}
+
+// MaxCapacity returns the maximum value to which Capacity can be set via ConnPool.SetCapacity
+func (pool *ConnPool[C]) MaxCapacity() int64 {
+	return pool.config.maxCapacity
+}
+
+// InUse returns the number of connections that the pool has lent out to clients and that
+// haven't been returned yet.
+func (pool *ConnPool[C]) InUse() int64 {
+	return pool.borrowed.Load()
+}
+
+// Available returns the number of connections that the pool can immediately lend out to
+// clients without blocking.
+func (pool *ConnPool[C]) Available() int64 {
+	return pool.capacity.Load() - pool.borrowed.Load()
+}
+
+// Active returns the numer of connections that the pool has currently open.
+func (pool *ConnPool[C]) Active() int64 {
+	return pool.active.Load()
+}
+
+func (pool *ConnPool[D]) IdleTimeout() time.Duration {
+	return time.Duration(pool.config.idleTimeout.Load())
+}
+
+func (pool *ConnPool[C]) SetIdleTimeout(duration time.Duration) {
+	pool.config.idleTimeout.Store(duration.Nanoseconds())
+}
+
+func (pool *ConnPool[D]) RefreshInterval() time.Duration {
+	return time.Duration(pool.config.refreshInterval.Load())
+}
+
+func (pool *ConnPool[C]) recordWait(start time.Time) {
+	pool.Metrics.waitCount.Add(1)
+	pool.Metrics.waitTime.Add(time.Since(start).Nanoseconds())
+	if pool.config.logWait != nil {
+		pool.config.logWait(start)
+	}
+}
+
+// Get returns a connection from the pool with the given Setting applied.
+// If there are no connections in the pool to be returned, Get blocks until one
+// is returned, or until the given ctx is cancelled.
+// The connection must be returned to the pool once it's not needed by calling Pooled.Recycle
+func (pool *ConnPool[C]) Get(ctx context.Context, setting *Setting) (*Pooled[C], error) {
+	if ctx.Err() != nil {
+		return nil, ErrCtxTimeout
+	}
+	if setting == nil {
+		return pool.get(ctx)
+	}
+	return pool.getWithSetting(ctx, setting)
+}
+
+// put returns a connection to the pool. This is a private API.
+// Return connections to the pool by calling Pooled.Recycle
+func (pool *ConnPool[C]) put(conn *Pooled[C]) {
+	pool.borrowed.Add(-1)
+
+	if conn != nil {
+		conn.timeUsed = time.Now()
+
+		lifetime := pool.extendedMaxLifetime()
+		if lifetime > 0 && time.Until(conn.timeCreated.Add(lifetime)) < 0 {
+			pool.Metrics.maxLifetimeClosed.Add(1)
+			conn.Close()
+			if err := pool.connReopen(context.Background(), conn, conn.timeUsed); err != nil {
+				return
+			}
+		}
+	}
+	if conn == nil {
+		var err error
+		conn, err = pool.connNew(context.Background())
```

Comment:
This follows the same pattern as the old pool: it doesn't use a context when re-opening because the MySQL client already fails fast, and you don't want to pass the request context to `Recycle` because it can often get cancelled (as we're at the end of the request when returning the connection) and cause the re-open to fail.